src/librustc_mir/borrow_check/nll/universal_regions.rs

```diff
@@ -490,6 +490,29 @@ impl<'cx, 'gcx, 'tcx> UniversalRegionsBuilder<'cx, 'gcx, 'tcx> {
             &bound_inputs_and_output,
             &mut indices,
         );
+
+        // Add late-bound regions which are not used in in inputs and outhout
+        let added_from_input_and_output = self.infcx.tcx.collect_referenced_late_bound_regions(
+            &bound_inputs_and_output);
+        let closure_base_def_id = self.infcx.tcx.closure_base_def_id(self.mir_def_id);
+        if let Some(late_bounds) = self.infcx.tcx.is_late_bound_map(closure_base_def_id.index) {
+            for late_bound in late_bounds.iter() {
+                let hir_id = HirId{ owner: closure_base_def_id.index, local_id: *late_bound };
+                let map = &self.infcx.tcx.hir;
+                let name = map.name(map.hir_to_node_id(hir_id)).as_interned_str();
+                let def_id = map.local_def_id(map.hir_to_node_id(hir_id));
+                let region = ty::BoundRegion::BrNamed(def_id, name);
+                if added_from_input_and_output.contains(&region) { continue; }
```

Comment:
Nit: I was thinking we could check `indices` to detect if the region already exists, but I guess in that case we'd have to create the `liberated_region` first. Still, we are doing that already below...

Reply:
Hmm, thinking a bit more about this, something seems a *touch* off. In particular, if we are looking at a closure, the `added_from_input_and_output` is going to contain late-bound regions from the **closure's** inputs and outputs. But here we are comparing against late-bound regions bound on the closure **owner** (`closure_base_def_id`).

I think comparing against entries in `indices` would be more robust. The diff would basically be to remove `added_from_input_and_output` and instead do:

```rust
let liberated_region = ...;
if !indices.contains_key(&liberated_region) {
    let region_vid = self.infcx.next_nll_region_var(FR);
    indices.insert_late_bound_region(liberated_region, region_vid.to_region_vid());
}
```